ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx

```diff
@@ -253,25 +303,48 @@ export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRe
     }
 
     private prepareGraph() {
-        const nodes = Object.values(this.props.nodes)
-            .filter(node => !!node)
-            .filter(node => node.phase !== NODE_PHASE.OMITTED)
-            .map(node => node.id);
+        const collapsedNodes: Set<string> = new Set();
+        const nodesToAdd: string[] = [];
         const edges = Object.values(this.props.nodes)
             .filter(node => !!node)
             .filter(node => node.phase !== NODE_PHASE.OMITTED)
-            .map(node =>
-                (node.children || [])
-                    // we can get outbound nodes, but no node
-                    .filter(childId => this.props.nodes[childId])
-                    .filter(childId => this.props.nodes[childId].phase !== NODE_PHASE.OMITTED)
-                    .map(childId => ({v: node.id, w: childId}))
-            )
+            .map(node => {
+                if (!node.children || node.children.length === 0) {
+                    return [];
+                } else if (node.children.length > 3 && !this.state.expandNodes.has('*') && !this.state.expandNodes.has(node.id)) {
+                    node.children.slice(1, node.children.length - 1).map(collapsedNode => collapsedNodes.add(collapsedNode));
```

Comment:
this loop looks like it could be O(n^2)

Reply:
Hmm, why do you say that? It seems to me it's `O(kn)` where `n` is the number of nodes and `k` is the average number of children per node. We can expect `k << n`, so it seems to be linear

Reply:
it is the map with a map that could be O(n^2), but you can easily prove me wrong by rendering a 2000 node graph